### PR TITLE
Internal priority queue

### PR DESCRIPTION
**Pull request by AnnaBlume99**

### CODE REVIEW

File: test/adiar/bdd/test_apply.cpp [hunk 1]
```diff
@@ -156,6 +166,8 @@ go_bandit([]() {
         AssertThat(out_nodes.can_pull(), Is().False());
 
         AssertThat(out.get<node_file>().meta_size(), Is().EqualTo(0u));
+
+        AssertThat((std::get<node_file>(out._union)._file_ptr)->max_1level_cut, Is().EqualTo(0u));
```
Comment (SSoelvsten) [inline test/adiar/bdd/test_apply.cpp:170]: I just realize, I actually have made a `.get<T>()` on the `__bdd` class such that this can be much neater:
```cpp
        AssertThat(out.get<node_file>()._file_ptr->max_1level_cut, Is().EqualTo(0u));
```
Sorry I just realize this now.

**Review by SSoelvsten — CHANGES_REQUESTED** [review]
The `std::get<arc_file>(...)` and `std::get<node_file>(...)` can actually be improved by using the lifted function `get<arc_file>()` and `get<node_file>()` on the `__bdd` and `__zdd` instances.

It looks good, except for that.

File: test/adiar/bdd/test_if_then_else.cpp [hunk 2]
```diff
@@ -1812,6 +1858,8 @@ go_bandit([]() {
       AssertThat(level_info.pull(), Is().EqualTo(create_level_info(0,1u)));
 
       AssertThat(level_info.can_pull(), Is().False());
+
+      AssertThat(out.get<node_file>()._file_ptr->max_1level_cut, Is().EqualTo(2u));
```
Comment (SSoelvsten) [inline test/adiar/bdd/test_if_then_else.cpp:1862]: The *zip* cases are implemented by use of creating the new output with the `<<` function in the `node_writer` (see the `ite_zip_bdds` function in *adiar/bdd/if_then_else.cpp*). Since `node_writer` does not take care of computing an over-approximation of the cuts (yet), then the default *max* value is still used.

The best (over-approximation) we can do for this edge case is
```cpp
std::max(bdd_if.max_1level_cut, bdd_then.max_1level_cut + bdd_else.max_1level_cut);
```
This does require us to do `.GreaterThanOrEqualTo(2u)`.

File: test/adiar/bdd/test_if_then_else.cpp [hunk 3]
```diff
@@ -1949,6 +1997,8 @@ go_bandit([]() {
...
```
Comment (SSoelvsten) [inline test/adiar/bdd/test_if_then_else.cpp:2001]: Just as above: add the over-approximation for the *zipping* case and turn this into a *greater or equals to*.

File: test/adiar/zdd/test_binop.cpp [hunk 4]
```diff
@@ -1,3 +1,5 @@
+#include <adiar/adiar.h>
```
Comment (SSoelvsten) [inline test/adiar/zdd/test_binop.cpp:1]: Does this really fix the issue on your machine, or does it just squelch it by erroring on this line alone? Does adding this help you? If not, remove.

File: test/adiar/zdd/test_build.cpp [hunk 5]
```diff
@@ -38,10 +42,12 @@ go_bandit([]() {
 
         level_info_test_stream<node_t> ms(res);
         AssertThat(ms.can_pull(), Is().False());
+
+        AssertThat(res.get<node_file>()._file_ptr->max_1level_cut, Is().GreaterThanOrEqualTo(0u));
```
Comment (SSoelvsten) [inline test/adiar/zdd/test_build.cpp:46]: This can be a `.EqualTo(0u)` since the construction can give the exact number.

File: test/adiar/zdd/test_build.cpp [hunk 6]
```diff
@@ -67,10 +75,12 @@ go_bandit([]() {
         AssertThat(ms.can_pull(), Is().True());
         AssertThat(ms.pull(), Is().EqualTo(create_level_info(0,1u)));
         AssertThat(ms.can_pull(), Is().False());
+
+        AssertThat(res.get<node_file>()._file_ptr->max_1level_cut, Is().GreaterThanOrEqualTo(0u));
```
Comment (SSoelvsten) [inline test/adiar/zdd/test_build.cpp:79]: This can be a `.EqualTo(0u)` since the construction can give the exact number.

File: test/adiar/zdd/test_build.cpp [hunk 7]
```diff
@@ -82,14 +92,16 @@ go_bandit([]() {
         AssertThat(ms.can_pull(), Is().True());
         AssertThat(ms.pull(), Is().EqualTo(create_level_info(42,1u)));
         AssertThat(ms.can_pull(), Is().False());
+
+        AssertThat(res.get<node_file>()._file_ptr->max_1level_cut, Is().GreaterThanOrEqualTo(0u));
```
Comment (SSoelvsten) [inline test/adiar/zdd/test_build.cpp:96]: This can be a `.EqualTo(0u)` since the construction can give the exact number.

File: test/adiar/zdd/test_build.cpp [hunk 8]
```diff
@@ -98,6 +110,8 @@ go_bandit([]() {
 
         level_info_test_stream<node_t> ms(res);
         AssertThat(ms.can_pull(), Is().False());
+
+        AssertThat(res.get<node_file>()._file_ptr->max_1level_cut, Is().GreaterThanOrEqualTo(0u));
```
Comment (SSoelvsten) [inline test/adiar/zdd/test_build.cpp:114]: This can be a `.EqualTo(0u)` since the construction can give the exact number.

File: test/adiar/zdd/test_build.cpp [hunk 9]
```diff
@@ -120,6 +134,8 @@ go_bandit([]() {
         AssertThat(ms.can_pull(), Is().True());
         AssertThat(ms.pull(), Is().EqualTo(create_level_info(42,1u)));
         AssertThat(ms.can_pull(), Is().False());
+
+        AssertThat(res.get<node_file>()._file_ptr->max_1level_cut, Is().GreaterThanOrEqualTo(0u));
```
Comment (SSoelvsten) [inline test/adiar/zdd/test_build.cpp:138]: This can be a `.EqualTo(0u)` since the construction can give the exact number.

File: test/adiar/zdd/test_build.cpp [hunk 10]
```diff
@@ -151,14 +167,16 @@ go_bandit([]() {
         AssertThat(ms.can_pull(), Is().True());
         AssertThat(ms.pull(), Is().EqualTo(create_level_info(1,1u)));
         AssertThat(ms.can_pull(), Is().False());
+
+        AssertThat(res.get<node_file>()._file_ptr->max_1level_cut, Is().GreaterThanOrEqualTo(1u));
```
Comment (SSoelvsten) [inline test/adiar/zdd/test_build.cpp:171]: This can be a `.EqualTo(1u)` since the construction can give the exact number.

File: test/adiar/zdd/test_build.cpp [hunk 11]
```diff
@@ -167,6 +185,8 @@ go_bandit([]() {
 
         level_info_test_stream<node_t> ms(res);
         AssertThat(ms.can_pull(), Is().False());
+
+        AssertThat(res.get<node_file>()._file_ptr->max_1level_cut, Is().GreaterThanOrEqualTo(0u));
```
Comment (SSoelvsten) [inline test/adiar/zdd/test_build.cpp:189]: This can be a `.EqualTo(0u)` since the construction can give the exact number.

File: test/adiar/zdd/test_build.cpp [hunk 12]
```diff
@@ -189,6 +209,8 @@ go_bandit([]() {
         AssertThat(ms.can_pull(), Is().True());
         AssertThat(ms.pull(), Is().EqualTo(create_level_info(42,1u)));
         AssertThat(ms.can_pull(), Is().False());
+
+        AssertThat(res.get<node_file>()._file_ptr->max_1level_cut, Is().GreaterThanOrEqualTo(0u));
```
Comment (SSoelvsten) [inline test/adiar/zdd/test_build.cpp:213]: This can be a `.EqualTo(0u)` since the construction can give the exact number.

File: test/adiar/zdd/test_build.cpp [hunk 13]
```diff
@@ -220,14 +242,16 @@ go_bandit([]() {
         AssertThat(ms.can_pull(), Is().True());
         AssertThat(ms.pull(), Is().EqualTo(create_level_info(1,1u)));
         AssertThat(ms.can_pull(), Is().False());
+
+        AssertThat(res.get<node_file>()._file_ptr->max_1level_cut, Is().GreaterThanOrEqualTo(1u));
```
Comment (SSoelvsten) [inline test/adiar/zdd/test_build.cpp:246]: This can be a `.EqualTo(1u)` since the construction can give the exact number.

File: test/adiar/zdd/test_build.cpp [hunk 14]
```diff
@@ -236,6 +260,8 @@ go_bandit([]() {
 
         level_info_test_stream<node_t> ms(res);
         AssertThat(ms.can_pull(), Is().False());
+
+        AssertThat(res.get<node_file>()._file_ptr->max_1level_cut, Is().GreaterThanOrEqualTo(0u));
```
Comment (SSoelvsten) [inline test/adiar/zdd/test_build.cpp:264]: This can be a `.EqualTo(0u)` since the construction can give the exact number.

File: test/adiar/zdd/test_build.cpp [hunk 15]
```diff
@@ -258,6 +284,8 @@ go_bandit([]() {
         AssertThat(ms.can_pull(), Is().True());
         AssertThat(ms.pull(), Is().EqualTo(create_level_info(42,1u)));
         AssertThat(ms.can_pull(), Is().False());
+
+        AssertThat(res.get<node_file>()._file_ptr->max_1level_cut, Is().GreaterThanOrEqualTo(0u));
```
Comment (SSoelvsten) [inline test/adiar/zdd/test_build.cpp:288]: This can be a `.EqualTo(0u)` since the construction can give the exact number.

File: test/adiar/zdd/test_build.cpp [hunk 16]
```diff
@@ -423,6 +463,8 @@ go_bandit([]() {
 
         level_info_test_stream<node_t> ms(res);
         AssertThat(ms.can_pull(), Is().False());
+
+        AssertThat(res.get<node_file>()._file_ptr->max_1level_cut, Is().GreaterThanOrEqualTo(0u));
```
Comment (SSoelvsten) [inline test/adiar/zdd/test_build.cpp:467]: This can be a `.EqualTo(0u)` since the construction can give the exact number.

File: test/adiar/zdd/test_build.cpp [hunk 17]
```diff
@@ -460,6 +502,8 @@ go_bandit([]() {
         AssertThat(ms.can_pull(), Is().True());
         AssertThat(ms.pull(), Is().EqualTo(create_level_info(0,1u)));
         AssertThat(ms.can_pull(), Is().False());
+
+        AssertThat(res.get<node_file>()._file_ptr->max_1level_cut, Is().GreaterThanOrEqualTo(2u));
```
Comment (SSoelvsten) [inline test/adiar/zdd/test_build.cpp:506]: This can be a `.EqualTo(2u)` since the construction can give the exact number.

File: test/adiar/zdd/test_build.cpp [hunk 18]
```diff
@@ -516,6 +562,8 @@ go_bandit([]() {
 
         level_info_test_stream<node_t> ms(res);
         AssertThat(ms.can_pull(), Is().False());
+
+        AssertThat(res.get<node_file>()._file_ptr->max_1level_cut, Is().GreaterThanOrEqualTo(0u));
```
Comment (SSoelvsten) [inline test/adiar/zdd/test_build.cpp:566]: This can be a `.EqualTo(0u)` since the construction can give the exact number.

File: test/adiar/zdd/test_build.cpp [hunk 19]
```diff
@@ -535,6 +583,8 @@ go_bandit([]() {
...
```
Comment (SSoelvsten) [inline test/adiar/zdd/test_build.cpp:587]: This can be a `.EqualTo(0u)` since the construction can give the exact number.

File: test/adiar/zdd/test_build.cpp [hunk 20]
```diff
@@ -581,6 +631,8 @@ go_bandit([]() {
         AssertThat(ms.can_pull(), Is().True());
         AssertThat(ms.pull(), Is().EqualTo(create_level_info(1,1u)));
         AssertThat(ms.can_pull(), Is().False());
+
+        AssertThat(res.get<node_file>()._file_ptr->max_1level_cut, Is().GreaterThanOrEqualTo(2u));
```
Comment (SSoelvsten) [inline test/adiar/zdd/test_build.cpp:635]: This can be a `.EqualTo(2u)` since the construction can give the exact number.

File: test/adiar/zdd/test_build.cpp [hunk 21]
```diff
@@ -648,6 +700,8 @@ go_bandit([]() {
         AssertThat(ms.can_pull(), Is().True());
         AssertThat(ms.pull(), Is().EqualTo(create_level_info(1,1u)));
         AssertThat(ms.can_pull(), Is().False());
+
+        AssertThat(res.get<node_file>()._file_ptr->max_1level_cut, Is().GreaterThanOrEqualTo(4u));
```
Comment (SSoelvsten) [inline test/adiar/zdd/test_build.cpp:704]: This can be a `.EqualTo(4u)` since the construction can give the exact number.

File: test/adiar/zdd/test_build.cpp [hunk 22]
```diff
@@ -797,6 +853,8 @@ go_bandit([]() {
         AssertThat(ms.can_pull(), Is().True());
         AssertThat(ms.pull(), Is().EqualTo(create_level_info(0,1u)));
         AssertThat(ms.can_pull(), Is().False());
+
+        AssertThat(res.get<node_file>()._file_ptr->max_1level_cut, Is().GreaterThanOrEqualTo(4u));
```
Comment (SSoelvsten) [inline test/adiar/zdd/test_build.cpp:857]: This can be a `.EqualTo(4u)` since the construction can give the exact number.

File: test/adiar/zdd/test_build.cpp [hunk 23]
```diff
@@ -923,6 +983,8 @@ go_bandit([]() {
         AssertThat(ms.can_pull(), Is().True());
         AssertThat(ms.pull(), Is().EqualTo(create_level_info(0,1u)));
         AssertThat(ms.can_pull(), Is().False());
+
+        AssertThat(res.get<node_file>()._file_ptr->max_1level_cut, Is().GreaterThanOrEqualTo(1u));
```
Comment (SSoelvsten) [inline test/adiar/zdd/test_build.cpp:987]: This can be a `.EqualTo(1u)` since the construction can give the exact number.

File: test/adiar/zdd/test_build.cpp [hunk 24]
```diff
@@ -977,6 +1039,8 @@ go_bandit([]() {
         AssertThat(ms.can_pull(), Is().True());
         AssertThat(ms.pull(), Is().EqualTo(create_level_info(0,1u)));
         AssertThat(ms.can_pull(), Is().False());
+
+        AssertThat(res.get<node_file>()._file_ptr->max_1level_cut, Is().GreaterThanOrEqualTo(3u));
```
Comment (SSoelvsten) [inline test/adiar/zdd/test_build.cpp:1043]: This can be a `.EqualTo(3u)` since the construction can give the exact number.

File: test/adiar/zdd/test_build.cpp [hunk 25]
```diff
@@ -1118,6 +1182,8 @@ go_bandit([]() {
         AssertThat(ms.can_pull(), Is().True());
         AssertThat(ms.pull(), Is().EqualTo(create_level_info(1,1u)));
         AssertThat(ms.can_pull(), Is().False());
+
+        AssertThat(res.get<node_file>()._file_ptr->max_1level_cut, Is().GreaterThanOrEqualTo(7u));
```
Comment (SSoelvsten) [inline test/adiar/zdd/test_build.cpp:1186]: This can be a `.EqualTo(7u)` since the construction can give the exact number.

File: test/adiar/zdd/test_build.cpp [hunk 26]
```diff
@@ -1377,6 +1445,8 @@ go_bandit([]() {
         AssertThat(ms.can_pull(), Is().True());
         AssertThat(ms.pull(), Is().EqualTo(create_level_info(0,1u)));
         AssertThat(ms.can_pull(), Is().False());
+
+        AssertThat(res.get<node_file>()._file_ptr->max_1level_cut, Is().GreaterThanOrEqualTo(12u));
```
Comment (SSoelvsten) [inline test/adiar/zdd/test_build.cpp:1449]: This can be a `.EqualTo(12u)` since the construction can give the exact number.

**Review by SSoelvsten — CHANGES_REQUESTED** [review]
Overall looks good. :slightly_smiling_face:  I have added comments that explain the reason and fixes for the last few issues in the tests.

File: src/adiar/zdd/build.cpp [hunk 27]
```diff
@@ -25,15 +25,16 @@ namespace adiar
...
+    /*adiar_assert(label <= MAX_LABEL, "Cannot represent that large a label");
 
     node_file nf;
     node_writer nw(nf);
     nw.unsafe_push(create_node(label, MAX_ID,
                                create_sink_ptr(false),
                                create_sink_ptr(true)));
     nw.unsafe_push(create_level_info(label,1u));
-    return nf;
+    return nf;*/
```
Comment (SSoelvsten) [inline src/adiar/zdd/build.cpp:36]: Since all of this logic is inside of `build_ithvar` this should not be commented out, but just removed.

**Review by SSoelvsten — COMMENTED** [review]
Tiny things left, I'll quickly do them.

**Review by SSoelvsten — APPROVED** [review]
